hathor/indexes/manager.py

```diff
@@ -66,23 +110,74 @@ def enable_tokens_index(self) -> None:
-        XXX: this method is not yet being used
+    def _manually_initialize(self, tx_storage: 'TransactionStorage') -> None:
+        """ Initialize the indexes, checking the indexes that need initialization, and the optimal iterator to use.
         """
-        for tx in tx_storage._topological_sort_timestamp_index():
-            tx_meta = tx.get_metadata()
-            if not tx_meta.validation.is_final():
-                continue
+        from hathor.transaction.genesis import BLOCK_GENESIS
+        from hathor.transaction.storage.transaction_storage import NULL_INDEX_LAST_STARTED_AT
 
-            self.all_tips.add_tx(tx)
+        db_last_started_at = tx_storage.get_last_started_at()
 
+        indexes_to_init: List[Tuple[_IndexFilter, BaseIndex]] = []
+        for index_filter, index in self._iter_all_indexes_with_filter():
+            index_db_name = index.get_db_name()
+            if index_db_name is None:
+                indexes_to_init.append((index_filter, index))
+                continue
+            index_last_started_at = tx_storage.get_index_last_started_at(index_db_name)
+            if db_last_started_at != index_last_started_at:
+                indexes_to_init.append((index_filter, index))
+
+        if indexes_to_init:
+            self.log.debug('there are indexes that need initialization',
+                           indexes_to_init=[i for _, i in indexes_to_init])
+        else:
+            self.log.debug('there are no indexes that need initialization')
+
+        # make sure that all the indexes that we're rebuilding are cleared
+        for _, index in indexes_to_init:
+            index_db_name = index.get_db_name()
+            if index_db_name:
+                tx_storage.set_index_last_started_at(index_db_name, NULL_INDEX_LAST_STARTED_AT)
+            index.force_clear()
+
+        block_count = 0
+        tx_count = 0
+        latest_timestamp = BLOCK_GENESIS.timestamp
+        first_timestamp = BLOCK_GENESIS.timestamp
+
+        for tx in progress(tx_storage.topological_iterator(), log=self.log):
+            # XXX: these would probably make more sense to be their own simple "indexes" instead of how it is here
+            latest_timestamp = max(tx.timestamp, latest_timestamp)
+            first_timestamp = min(tx.timestamp, first_timestamp)
             if tx.is_block:
-                self.block_tips.add_tx(tx)
-            elif not tx_meta.voided_by:
-                self.tx_tips.add_tx(tx)
+                block_count += 1
+            else:
+                tx_count += 1
+
+            tx_meta = tx.get_metadata()
+
+            # feed each transaction to the indexes that they are interested in
+            for index_filter, index in indexes_to_init:
+                if index_filter is _IndexFilter.ALL:
+                    index.init_loop_step(tx)
+                elif index_filter is _IndexFilter.VALID_BLOCKS:
+                    if tx.is_block:
```

Comment:
I was doing that initially, but then realized that the indexes that use that filter actually want the voided blocks, so maybe I should rename the filter.

Reply:
Sounds good.

Reply:
There is actually a problem with this line. The neither filter name is completely accurate, because neither excluding all voided blocks nor including all voided blocks will _necessarily_ rebuild the index to the same state it was before.

The problematic case is when there are multiple valid heads: they are voided _and_ tips. Which makes it so they are not removed from the block tips index when they are marked as voided, but they are marked as voided anyway.

I'm not completely sure that the workaround should be. I think this is just one more case of a situation that could have been avoided if we never allowed ties. However that change itself must be done more carefully, but I really think we should consider it and prioritize it soon, it's probably possible to do it in a completely backwards way.

We also probably have that bug already, the current initialization does not deal with this correctly (currently we add everything to the indexes and immediately remove what's voided, the remove function will skip the indexes that should not be affected when `remove_all=False`, but that's it).

For now I prefer to have some sort of workaround to make sure the index is restored to the desired state, although I'm not really sure what's best. I tested using this condition: `tx.is_block and (not tx_meta.voided_by or not tx_meta.children)`, it works for all test cases, but it's still wrong because old voided tips that aren't currently heads will end up as block tips.

Reply:
From what we talked the de-facto implementation _is_ to add all blocks, voided or not, to the index. But there's also some small inconsistency in how add and remove blocks to the blocks_tips, but this can be dealt separately.